### PR TITLE
prov/psm2: Add scalable endpoint support

### PR DESCRIPTION
**Pull request by j-xiong**
This set of commits implement scalable endpoints in the psm2 provider. It allows applications to utilize multiple hardware Tx/Rx contexts to achieve higher bandwidth over a single endpoint.

### CODE REVIEW

**Comment by j-xiong** [comment]
@shefty The changes are big, but all confined within the psm2 provider.

**Comment by shefty** [comment]
@j-xiong - Do you want me to do a detailed review of the changes?

**Comment by j-xiong** [comment]
@shefty Any level of review is good. Depending on your time.

File: prov/psm2/src/psmx2.h [hunk 1]
```diff
@@ -723,6 +738,27 @@ struct psmx2_fid_ep {
...
+	/* above fields are common with regular ep */
+
+	struct dlist_entry	entry;
+
+	ofi_atomic32_t		ref;
+	int			service;
+	uint8_t			id;
+	uint8_t			enabled;
+	size_t			ctxt_cnt;
+	struct psmx2_sep_ctxt	ctxts[0]; /* must be last element */
```
Comment (shefty) [inline prov/psm2/src/psmx2.h]: I think the standard is to use [ ] without the 0.

File: prov/psm2/src/psmx2.h [hunk 2]
```diff
@@ -690,18 +709,31 @@ struct psmx2_fid_cntr {
...
 };
 
+struct psmx2_ctxt_addr {
+	psm2_epid_t		epid;
+	psm2_epaddr_t		epaddr;
+};
+
+struct psmx2_sep_addr {
+	int			ctxt_cnt;
+	struct psmx2_ctxt_addr	ctxt_addrs[0];
```
Comment (shefty) [inline prov/psm2/src/psmx2.h]: Can use [ ]

File: prov/psm2/src/psmx2_av.c [hunk 3]
```diff
@@ -32,6 +32,137 @@
...
+	struct psmx2_fid_sep *sep;
+
+	sep = container_of(entry, struct psmx2_fid_sep, entry);
+	return ((uintptr_t)sep->id == (uintptr_t)arg);
+}
+
+static void psmx2_am_sep_completion(void *buf)
+{
+	if (buf)
+		free(buf);
```
Comment (shefty) [inline prov/psm2/src/psmx2_av.c]: free can take NULL as input

File: prov/psm2/src/psmx2_av.c [hunk 4]
```diff
@@ -32,6 +32,137 @@
...
+		if (!entry) {
+			op_error = PSM2_EPID_UNKNOWN;
+			n = 0;
+			buflen = 0;
+		} else {
+			sep = container_of(entry, struct psmx2_fid_sep, entry);
+			n = sep->ctxt_cnt;
+			buflen = n * sizeof(psm2_epid_t);
+			if (n) {
+				buf = malloc(buflen);
```
Comment (shefty) [inline prov/psm2/src/psmx2_av.c]: need to check for malloc failure

File: prov/psm2/src/psmx2_av.c [hunk 5]
```diff
@@ -32,6 +32,137 @@
...
+			p  = calloc(1, sizeof (struct psmx2_sep_addr) +
+				       n * sizeof(struct psmx2_ctxt_addr));
+			if (!p) {
+				ofi_atomic_inc32(&req->error_count);
+				req->errors[i] = PSM2_NO_MEMORY;
+			}
+			p->ctxt_cnt = n;
+			for (j=0; j<n; j++) {
+				p->ctxt_addrs[j].epid = ((psm2_epid_t *)src)[j];
+				p->ctxt_addrs[j].epaddr = (psm2_epaddr_t)FI_ADDR_NOTAVAIL;
```
Comment (shefty) [inline prov/psm2/src/psmx2_av.c]: p may be NULL if calloc fails

File: prov/psm2/src/psmx2_av.c [hunk 6]
```diff
@@ -223,38 +223,50 @@ int psmx2_epid_to_epaddr(struct psmx2_fid_domain *domain,
...
+		err = psmx2_epid_to_epaddr(trx_ctxt,
 					   av->sepaddrs[idx]->ctxt_addrs[ctxt].epid,
 					   &epaddr);
-		if (!err)
-			av->sepaddrs[idx]->ctxt_addrs[ctxt].epaddr = epaddr;
+		if (err) {
+			FI_WARN(&psmx2_prov, FI_LOG_AV,
+				"fatal error: unable to translate epid %lx to epaddr.\n",
+				av->sepaddrs[idx]->ctxt_addrs[ctxt].epid);
+			exit(-1);
```
Comment (shefty) [inline prov/psm2/src/psmx2_av.c]: calling exit from a library is rather rude :) Maybe assert() and return NULL?

File: prov/psm2/src/psmx2_tagged.c [hunk 7]
```diff
@@ -996,7 +1065,10 @@ ssize_t psmx2_tagged_sendv_generic(struct fid_ep *ep,
 	}
 
 	av = ep_priv->av;
-	if (av && av->type == FI_AV_TABLE) {
+	if (av && PSMX2_SEP_ADDR_TEST(dest_addr)) {
+		psm2_epaddr = psmx2_av_translate_sep(av, ep_priv->trx_ctxt, dest_addr);
+		vlane = 0;
+	} else  if (av && av->type == FI_AV_TABLE) {
```
Comment (shefty) [inline prov/psm2/src/psmx2_tagged.c]: Is there some way to encapsulate these changes into a call?  This same set of changes is repeated about a dozen times or more.

File: prov/psm2/src/psmx2_init.c [hunk 8]
```diff
@@ -62,6 +65,23 @@ static void psmx2_init_env(void)
...
+	uint64_t (*func)(uint64_t);
+	int mask = 1; /* PSM2_MULTI_EP_CAP */
+
+	func = dlsym(NULL, "psm2_get_capability_mask");
+	if (!func)
+		return 0;
+
+	if ((*func)(mask) != mask)
+		return 0;
+
```
Comment (shefty) [inline prov/psm2/src/psmx2_init.c]: Need dlclose

File: prov/psm2/src/psmx2_init.c [hunk 9]
```diff
@@ -122,6 +120,53 @@ static int psmx2_init_lib(void)
...
+	n = read(fd, buffer, 32);
+	if (n > 0) {
+		buffer[n] = 0;
+		ret = strtol(buffer, NULL, 10);
+		FI_INFO(&psmx2_prov, FI_LOG_CORE, "%s: %d\n", path, ret);
+	}
+	close(fd);
+
+	return ret;
+}
```
Comment (shefty) [inline prov/psm2/src/psmx2_init.c]: I think you can use fi_read_file() -- see fi_file.h

**Comment by shefty** [comment]
I reviewed this by examining each commit individually, so not all comments apply to the final set of changes.

**Comment by j-xiong** [comment]
@shefty Thanks! Will address the issues and update the PR.

File: prov/psm2/src/psmx2_tagged.c (hunk 7 above)
Comment (j-xiong) [inline prov/psm2/src/psmx2_tagged.c]: I think it would be better handled via a separate PR. The consolidation will be applied not only to the newly added code, but also the existing code in the other conditional branches.

**Comment by j-xiong** [comment]
@shefty The commits are updated. I have addressed all the comments except for the SEP address testing (will optimize in a future PR) and the missing of dlclose.

**Comment by shefty** [comment]
👍 thanks

**Comment by rzambre** [comment]
When I to use it with opa-psm2, my `domain->ops->scalable_ep` gets initialized with `fi_no_scalable_ep`. That's because `psmx2_sep_ok` returns 0 if `PSM2_MULTI_EP_CAP` is not defined. But I don't see `PSM2_MULTI_EP_CAP` in the `01org/opa-psm2` repo.

1) Will this flag be defined in a future release of PSM2?
2) Is there any way to use this feature with `01org/opa-psm2`? Hacking out the `#ifdef` won't work because `psm2_get_capability_mask`, `psm2_ep_epid_lookup2` haven't made it upstream to opa-psm2.

**Comment by j-xiong** [comment]
1. yes
2. you have to wait until the new feature gets upstream.

I agree this can be confusing but PSM2 and libfabric have different development cycle. We have to get the provider update before PSM2 in order to fit into the 1.5 release time frame. Although it has been tested internally, right now please consider the scalable endpoint support experimental.

**Comment by rzambre** [comment]
Thank you! Any estimate on when the new feature will make it upstream into `opa-psm2`?

**Comment by j-xiong** [comment]
I don't have the information.

**Comment by rzambre** [comment]
Could you point me to an example-case that uses this? Or a development branch of the newer psm2? We are running some experiments here and so, we don't need the production version of the newer PSM2.

**Comment by j-xiong** [comment]
Here is an example: https://github.com/j-xiong/ofi-test.

Unfortunately I can't give you a development branch of the newer psm2. You have to wait for that being pushed to github.
